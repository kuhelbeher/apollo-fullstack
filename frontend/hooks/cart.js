import { useQuery, useMutation } from 'react-apollo';
import gql from 'graphql-tag';

export const LOCAL_STATE_QUERY = gql`
  query {
    cartOpen @client
  }
`;

export const TOGGLE_CART_MUTATION = gql`
  mutation {
    toggleCart @client
  }
`;

export const useCart = () => {
  const {
    data: { cartOpen },
  } = useQuery(LOCAL_STATE_QUERY);

  const [toggleCart] = useMutation(TOGGLE_CART_MUTATION);

  return [cartOpen, toggleCart];
};
