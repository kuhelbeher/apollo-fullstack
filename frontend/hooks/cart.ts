import { useQuery, useMutation } from 'react-apollo';
import gql from 'graphql-tag';

type LocalStateData = {
  cartOpen: boolean;
};

export const LOCAL_STATE_QUERY = gql`
  query {
    cartOpen @client
  }
`;

export const TOGGLE_CART_MUTATION = gql`
  mutation {
    toggleCart @client
  }
`;

export const useCart = (): [boolean | undefined, () => void] => {
  const { data: { cartOpen } = {} } = useQuery<LocalStateData>(
    LOCAL_STATE_QUERY
  );

  const [toggleCart] = useMutation(TOGGLE_CART_MUTATION);

  return [cartOpen, toggleCart];
};
