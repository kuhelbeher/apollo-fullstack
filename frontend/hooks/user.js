import { useQuery } from 'react-apollo';
import gql from 'graphql-tag';

export const CURRENT_USER_QUERY = gql`
  query CURRENT_USER_QUERY {
    me {
      id
      email
      name
      permissions
    }
  }
`;

export const useUser = () => {
  const payload = useQuery(CURRENT_USER_QUERY);

  return payload;
};
