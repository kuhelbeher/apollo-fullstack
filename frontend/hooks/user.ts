import { useQuery } from 'react-apollo';
import gql from 'graphql-tag';

import { Permissions, CartItem } from '../types';

export type CurrentUserType = {
  me: {
    id: string;
    email: string;
    name: string;
    permissions: Permissions;
    cart: CartItem[];
  };
};

export const CURRENT_USER_QUERY = gql`
  query CURRENT_USER_QUERY {
    me {
      id
      email
      name
      permissions
      cart {
        id
        quantity
        item {
          id
          title
          price
          image
          description
        }
      }
    }
  }
`;

export const useUser = () => {
  const payload = useQuery<CurrentUserType>(CURRENT_USER_QUERY);

  return payload;
};
