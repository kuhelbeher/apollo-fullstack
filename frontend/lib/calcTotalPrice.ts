import { CartItem } from '../types';

export default function calcTotalPrice(cart: CartItem[]): number {
  return cart.reduce((tally, cartItem) => {
    if (!cartItem.item) return tally;
    return tally + cartItem.quantity * cartItem.item.price;
  }, 0);
}
