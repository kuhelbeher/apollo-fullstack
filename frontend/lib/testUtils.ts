import casual from 'casual';
import { act } from 'react-dom/test-utils';
import { wait } from '@apollo/react-testing';
import { ReactWrapper } from 'enzyme';

// seed it so we get consistent results
casual.seed(777);

const fakeItem = () => ({
  __typename: 'Item',
  id: 'abc123',
  price: 5000,
  user: null,
  image: 'dog-small.jpg',
  title: 'dogs are best',
  description: 'dogs',
  largeImage: 'dog.jpg',
});

const fakeUser = () => ({
  __typename: 'User',
  id: '4234',
  name: casual.name,
  email: casual.email,
  permissions: ['ADMIN'],
  cart: [],
});

const fakeOrderItem = () => ({
  __typename: 'OrderItem',
  id: casual.uuid,
  image: `${casual.word}.jpg`,
  title: casual.words(),
  price: 4234,
  quantity: 1,
  description: casual.words(),
});

const fakeOrder = () => ({
  __typename: 'Order',
  id: 'ord123',
  charge: 'ch_123',
  total: 40000,
  items: [fakeOrderItem(), fakeOrderItem()],
  createdAt: '2018-04 - 06T19: 24: 16.000Z',
  user: fakeUser(),
});

type Overrides = {
  id: string;
  quantity: number;
};

const fakeCartItem = (overrides: Overrides) => ({
  __typename: 'CartItem',
  id: 'omg123',
  quantity: 3,
  item: fakeItem(),
  user: fakeUser(),
  ...overrides,
});

// Fake LocalStorage
class LocalStorageMock {
  store: { [key: string]: string } = {};

  clear() {
    this.store = {};
  }

  getItem(key: string) {
    return this.store[key] || null;
  }

  setItem(key: string, value: any) {
    this.store[key] = value.toString();
  }

  removeItem(key: string) {
    delete this.store[key];
  }
}

// Use this in your test after mounting if you need just need to let the query finish without updating the wrapper
async function actWait(amount = 0) {
  await act(async () => {
    await wait(amount);
  });
}

// Use this in your test after mounting if you want the query to finish and update the wrapper
async function updateWrapper(wrapper: ReactWrapper, amount = 0) {
  await act(async () => {
    await wait(amount);
    wrapper.update();
  });
}

export {
  LocalStorageMock,
  fakeItem,
  fakeUser,
  fakeCartItem,
  fakeOrder,
  fakeOrderItem,
  actWait,
  updateWrapper,
};
