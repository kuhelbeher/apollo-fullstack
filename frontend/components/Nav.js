import React from 'react';
import Link from 'next/link';

import NavStyles from './styles/NavStyles';
import { useUser, useCart } from '../hooks';
import Signout from './Signout';
import CartCount from './CartCount';

function Nav() {
  const {
    data: { me },
  } = useUser();
  const [, toggleCart] = useCart();

  return (
    <NavStyles data-test="nav">
      <Link href="/items">
        <a>Shop</a>
      </Link>
      {me && (
        <>
          <Link href="/sell">
            <a>Sell</a>
          </Link>
          <Link href="/orders">
            <a>Orders</a>
          </Link>
          <Link href="/me">
            <a>Account</a>
          </Link>
          <Signout />
          <button type="button" onClick={toggleCart}>
            My Cart
            <CartCount
              count={me.cart.reduce(
                (tally, cartItem) => tally + cartItem.quantity,
                0
              )}
            />
          </button>
        </>
      )}
      {!me && (
        <Link href="/signup">
          <a>Sign In</a>
        </Link>
      )}
    </NavStyles>
  );
}

export default Nav;
