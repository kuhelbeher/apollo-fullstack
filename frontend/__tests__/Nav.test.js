import { mount } from 'enzyme';
import { MockedProvider } from '@apollo/react-testing';
import toJSON from 'enzyme-to-json';

import Nav from '../components/Nav';
import { CURRENT_USER_QUERY, LOCAL_STATE_QUERY } from '../hooks';
import { fakeUser, updateWrapper, fakeCartItem } from '../lib/testUtils';

const notSignedInMocks = [
  { request: { query: CURRENT_USER_QUERY }, result: { data: { me: null } } },
  {
    request: { query: LOCAL_STATE_QUERY },
    result: { data: { cartOpen: false } },
  },
];

const signedInMocks = [
  {
    request: { query: CURRENT_USER_QUERY },
    result: { data: { me: fakeUser() } },
  },
  {
    request: { query: LOCAL_STATE_QUERY },
    result: { data: { cartOpen: false } },
  },
];
const signedInMocksWithCartItems = [
  {
    request: { query: CURRENT_USER_QUERY },
    result: {
      data: {
        me: {
          ...fakeUser(),
          cart: [fakeCartItem(), fakeCartItem(), fakeCartItem()],
        },
      },
    },
  },
  {
    request: { query: LOCAL_STATE_QUERY },
    result: { data: { cartOpen: false } },
  },
];

describe('<Nav />', () => {
  it('renders a minimal nav when signed out', async () => {
    const wrapper = mount(
      <MockedProvider mocks={notSignedInMocks}>
        <Nav />
      </MockedProvider>
    );

    await updateWrapper(wrapper);

    const nav = wrapper.find('ul[data-test="nav"]');

    expect(toJSON(nav)).toMatchSnapshot();
  });

  it('renders a full nav when signin', async () => {
    const wrapper = mount(
      <MockedProvider mocks={signedInMocks}>
        <Nav />
      </MockedProvider>
    );

    await updateWrapper(wrapper);

    const nav = wrapper.find('ul[data-test="nav"]');

    expect(nav.children()).toHaveLength(6);
    expect(nav.text()).toContain('Sign Out');

    // expect(toJSON(nav)).toMatchSnapshot();
  });

  it('renders the amount of items in the cart', async () => {
    const wrapper = mount(
      <MockedProvider mocks={signedInMocksWithCartItems}>
        <Nav />
      </MockedProvider>
    );

    await updateWrapper(wrapper);

    const nav = wrapper.find('ul[data-test="nav"]');
    const count = nav.find('div.count');

    expect(toJSON(count)).toMatchSnapshot();
  });
});
