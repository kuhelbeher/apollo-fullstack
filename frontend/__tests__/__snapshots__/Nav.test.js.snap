// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Nav /> renders a minimal nav when signed out 1`] = `
<ul
  className="NavStyles-sc-11c0d2g-0 dDiqeM"
  data-test="nav"
>
  <Link
    href="/items"
    prefetch={true}
  >
    <a
      href="/items"
      onClick={[Function]}
      onMouseEnter={[Function]}
    >
      Shop
    </a>
  </Link>
  <Link
    href="/signup"
    prefetch={true}
  >
    <a
      href="/signup"
      onClick={[Function]}
      onMouseEnter={[Function]}
    >
      Sign In
    </a>
  </Link>
</ul>
`;

exports[`<Nav /> renders the amount of items in the cart 1`] = `
<div
  className="CartCount__Dot-xxvp4g-1 fJsVOg count"
>
  9
</div>
`;
